client/go/outline/config/config_iptable.go

```diff
@@ -0,0 +1,162 @@
+)
+
+type ipTableEntryConfig struct {
+	IP     string                `yaml:"ip,omitempty"`
+	Dialer configyaml.ConfigNode `yaml:"dialer"`
+}
+
+type ipTableRootConfig struct {
+	Table []ipTableEntryConfig `yaml:"table"`
+}
+
+// parsedIPTableStreamEntry is an internal struct to hold a parsed prefix and its corresponding stream dialer.
+type parsedIPTableStreamEntry struct {
+	prefix netip.Prefix
+	dialer transport.StreamDialer
+}
+
+func parseIPTableStreamDialer(
+	ctx context.Context,
+	configMap map[string]any,
+	parseSD configyaml.ParseFunc[*Dialer[transport.StreamConn]],
+) (*iptable.StreamDialer, bool, error) {
+	var rootCfg ipTableRootConfig
+	if err := configyaml.MapToAny(configMap, &rootCfg); err != nil {
+		return nil, false, fmt.Errorf("failed to map iptable stream config: %w", err)
+	}
+
+	if len(rootCfg.Table) == 0 {
+		return nil, false, errors.New("iptable config 'table' must not be empty for stream dialer")
+	}
+
+	parsedEntries := make([]parsedIPTableStreamEntry, 0, len(rootCfg.Table))
+	var defaultDialerEntry *parsedIPTableStreamEntry
+	hasDirectConn := false
+
+	for i, entryCfg := range rootCfg.Table {
+		parsedSubDialer, err := parseSD(ctx, entryCfg.Dialer)
+		if err != nil {
+			return nil, false, fmt.Errorf("failed to parse nested stream dialer for table entry %d (ip: %s): %w", i, entryCfg.IP, err)
+		}
+
+		if parsedSubDialer.ConnType == ConnTypeDirect {
+			hasDirectConn = true
+		}
+
+		currentEntry := parsedIPTableStreamEntry{
+			dialer: transport.FuncStreamDialer(parsedSubDialer.Dial),
+		}
+
+		if entryCfg.IP == "" { // Default dialer
+			if defaultDialerEntry != nil {
+				return nil, false, errors.New("multiple default dialers specified in iptable for stream")
+			}
+			defaultDialerEntry = &currentEntry
+			continue
+		}
+
+		var prefix netip.Prefix
+		parsedPrefix, errPrefix := netip.ParsePrefix(entryCfg.IP)
+		if errPrefix == nil {
+			prefix = parsedPrefix
+		} else {
+			addr, errAddr := netip.ParseAddr(entryCfg.IP)
+			if errAddr != nil {
+				return nil, false, fmt.Errorf("iptable entry %d IP '%s' is not a valid IP address or CIDR prefix: failed to parse as prefix (%v) and failed to parse as address (%v)", i, entryCfg.IP, errPrefix, errAddr)
+			}
+			prefix = netip.PrefixFrom(addr, addr.BitLen())
+		}
+
+		currentEntry.prefix = prefix
+		parsedEntries = append(parsedEntries, currentEntry)
+	}
+
+	table := iptable.NewIPTable[transport.StreamDialer]()
+
+	for _, entry := range parsedEntries {
+		table.AddPrefix(entry.prefix, entry.dialer)
+	}
+
+	dialer, err := iptable.NewStreamDialer(table)
+
+	if err != nil {
+		return nil, false, fmt.Errorf("failed to create IPTableStreamDialer: %w", err)
+	}
+
+	isFullTunnel := !hasDirectConn && (defaultDialerEntry != nil || isExhaustive(parsedEntries))
+	if defaultDialerEntry != nil {
+		dialer.SetDefault(defaultDialerEntry.dialer)
+	}
+
+	return dialer, isFullTunnel, nil
+}
+
+func isExhaustive(entries []parsedIPTableStreamEntry) bool {
```

Comment:
This is not needed, and I don't think it's correct.

The table is always exhaustive if there's a default dialer or empty entry (perhaps we should ditch the default dialer, it's redundant)

If it's not exhaustive, it doesn't matter, because we only care about the routes that dialer is configured for.

Reply:
The ConnType for the parent dialer depends on the ConnType of the childred. One entry doesn't mean Tunneled. It means the ConnType of that one child dialer.

Reply:
This is tunneled, because all the possible dialers are tunneled.

Reply:
If all the possible dialers were direct, then it would be direct